### PR TITLE
Agreements api v2 [WIP]

### PR DESCRIPTION
**Pull request by ilcsik**

### CODE REVIEW

File: EquiTrack/partners/views_v2.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            if "partner_name" in query_params.keys():
+                queries.append(Q(partner__name__icontains=query_params.get("partner_name")))
+            if "start" in query_params.keys():
+                queries.append(Q(start__gt=query_params.get("start")))
+            if "end" in query_params.keys():
+                queries.append(Q(end__lte=query_params.get("end")))
+            if "search" in query_params.keys():
+                queries.append(Q(partner__name__icontains=query_params.get("search")))
+
+            expression = functools.reduce(operator.and_, queries)
```
Comment (ntrncic) [inline EquiTrack/partners/views_v2.py]: why is this needed?

File: EquiTrack/partners/views_v2.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+from rest_framework.decorators import detail_route
+
+from .models import Agreement, PCA
+from .serializers import AgreementSerializer
+from .serializers_v2 import AgreementListSerializer
+from .permissions import PartnerPermission
+from .filters import PartnerScopeFilter
+
+
+class AgreementListAPIView(ListCreateAPIView):
```
Comment (ntrncic) [inline EquiTrack/partners/views_v2.py]: we also need a patch request so maybe extend RetrieveUpdateAPIView as well\
we need another method for patch (edit)

File: EquiTrack/partners/views_v2.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        else:
+            return super(AgreementListAPIView, self).get_serializer_class()
+
+    def create(self, request, *args, **kwargs):
+        """
+        Add a new Agreement
+        :return: JSON
+        """
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
```
Comment (ntrncic) [inline EquiTrack/partners/views_v2.py]: please add field validations. they are found in the forms and in the Agreements tab of the Requirements spreadsheet.

File: EquiTrack/partners/views_v2.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            return super(AgreementListAPIView, self).get_queryset()
+
+
+class AgreementDetailAPIView(RetrieveUpdateAPIView):
+    """
+    Retrieve and Update Agreement.
+    """
+    queryset = Agreement.objects.all()
+    serializer_class = AgreementSerializer
+    permission_classes = (PartnerPermission,)
```
Comment (ntrncic) [inline EquiTrack/partners/views_v2.py]: check permissions as shown in the requirements document. only partnership manager can add and update.

File: EquiTrack/partners/views_v2.py (hunk 2 above)
Comment (ilcsik) [inline EquiTrack/partners/views_v2.py]: PATCH is handled by AgreementDetailAPIView which is a RetrieveUpdateAPIView. There's a test for it. This view is only handles listing and creation. This is to address what we discussed regarding moving away from ViewSets, Rob+Simon kind of agreed to use this practice (separate view for create/list, and retrieve/update/delete due to the GET method routing)

File: EquiTrack/partners/views_v2.py (hunk 3 above)
Comment (ilcsik) [inline EquiTrack/partners/views_v2.py]: I'm working on it, I was kind of stuck but commiting soon.

File: EquiTrack/partners/views_v2.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        if query_params:
+            queries = []
+
+            if "agreement_type" in query_params.keys():
+                queries.append(Q(agreement_type=query_params.get("agreement_type")))
+            if "status" in query_params.keys():
+                queries.append(Q(status=query_params.get("status")))
+            if "partner_name" in query_params.keys():
+                queries.append(Q(partner__name__icontains=query_params.get("partner_name")))
+            if "start" in query_params.keys():
```
Comment (ilcsik) [inline EquiTrack/partners/views_v2.py]: Yes, if it's a dropdown I'll refactor, thanks!

File: EquiTrack/partners/views_v2.py (hunk 4 above)
Comment (ilcsik) [inline EquiTrack/partners/views_v2.py]: Yep, thanks!

File: EquiTrack/partners/serializers_v2.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    class Meta:
+        model = Agreement
+        fields = (
+            "id",
+            "reference_number",
+            "partner_name",
+            "agreement_type",
+            "end",
+            "start",
+            "signed_by_unicef_date",
```
Comment (ilcsik) [inline EquiTrack/partners/serializers_v2.py]: I used the Field mapping as a reference, 2.5.1.1 - 2.5.1.9 (Agreements list/Display). Attachment is indeed missing, but the other two is not mentioned there. Should I add those?

File: EquiTrack/partners/urls_v2.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+from .views_v2 import (
+    AgreementListAPIView,
+    AgreementDetailAPIView,
+)
+
+urlpatterns = (
+    url(r'^partners/(?P<partner_pk>\d+)/agreements/$', view=AgreementListAPIView.as_view(), name='parter-agreement-list'),
+    url(r'^partners/(?P<partner_pk>\d+)/agreements/(?P<pk>\d+)/$', view=AgreementDetailAPIView.as_view(), name='partner-agreement-detail'),
+)
```
Comment (ilcsik) [inline EquiTrack/partners/urls_v2.py]: Ok.

**Comment by ilcsik** [comment]
What do you mean by create packages for partners app?

File: EquiTrack/partners/views_v2.py (hunk 1 above)
Comment (ntrncic) [inline EquiTrack/partners/views_v2.py]: ok that's cool!

**Comment by ilcsik** [comment]
I have one question regarding reference_number, which is a property method currently. As per the sheet it needs to be moved to the save method and calculate/save only once when an Agreement goes active, right? My question is, are calculation rules the same so I can just copy the current implementation? I don't have access to the doc included in the sheet that explains.

File: EquiTrack/partners/serializers_v2.py (hunk 6 above)
Comment (ntrncic) [inline EquiTrack/partners/serializers_v2.py]: yes add them.

File: EquiTrack/partners/views_v2.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            status=status.HTTP_201_CREATED,
+            headers=headers
+        )
+
+    @detail_route(methods=['get'], url_path='interventions')
+    def interventions(self, request, partner_pk=None, pk =None):
+        """
+        Return All Interventions for Partner and Agreement
+        """
+        data = PCA.objects.filter(partner_id=partner_pk, agreement_id=pk).values()
```
Comment (robertavram) [inline EquiTrack/partners/views_v2.py]: this list of objects need to be properly serialized..

Comment (ilcsik) [inline EquiTrack/partners/views_v2.py]: Indeed (this was copied from the original view), refactored in https://github.com/unicef/etools/pull/225/commits/e1bb347992520b35a5d6099e41b19235de5d1bf8

File: EquiTrack/partners/models.py [hunk 9]
```diff
@@ -619,8 +619,6 @@ class Agreement(TimeStampedModel):
         (PCA, u"Programme Cooperation Agreement"),
         (SSFA, u'Small Scale Funding Agreement'),
         (MOU, u'Memorandum of Understanding'),
-        (IC, u'Institutional Contract'),
```
Comment (ilcsik) [inline EquiTrack/partners/models.py]: Fixed in https://github.com/unicef/etools/pull/225/commits/4f8523135108c24d5dd81c9607823705258278a1

File: EquiTrack/locations/migrations/0008_auto_20161109_1059.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('locations', '0007_auto_20161027_2022'),
+    ]
+
+    operations = [
+        migrations.AlterModelManagers(
```
Comment (ilcsik) [inline EquiTrack/locations/migrations/0008_auto_20161109_1059.py]: Hmm it was generated by makemigrations after I updated my branch from develop. Not sure what's going on, removed in https://github.com/unicef/etools/pull/225/commits/7cb0135c65cb4408da87e6cb85c4b1e4095249ef

File: EquiTrack/partners/urls_v2.py (hunk 7 above)
Comment (ilcsik) [inline EquiTrack/partners/urls_v2.py]: Done in https://github.com/unicef/etools/pull/225/commits/bba55b0843285f9e3bdd851f9db8f237d8fc575e

File: EquiTrack/partners/serializers_v2.py (hunk 6 above)
Comment (ilcsik) [inline EquiTrack/partners/serializers_v2.py]: Done in https://github.com/unicef/etools/pull/225/commits/af5a66d7129c61cf28a25fee3226d132d83e1517

File: EquiTrack/partners/views_v2.py (hunk 4 above)
Comment (ilcsik) [inline EquiTrack/partners/views_v2.py]: Permissions added in https://github.com/unicef/etools/pull/225/commits/2e06ee57e061f1f6b121d650ba6118b40f035c5e

Please review I'm not sure if this covers what I see in the sheet? Also I introduced a new permission class rather than refactor the old, so that old endpoints won't break.

File: EquiTrack/partners/views_v2.py (hunk 5 above)
Comment (ilcsik) [inline EquiTrack/partners/views_v2.py]: Done in https://github.com/unicef/etools/pull/225/commits/2eed29e6a5bd9d368c5b9817cbf458925114923e

File: EquiTrack/partners/views_v2.py (hunk 2 above)
Comment (ilcsik) [inline EquiTrack/partners/views_v2.py]: Done in https://github.com/unicef/etools/pull/225/commits/bba55b0843285f9e3bdd851f9db8f237d8fc575e

**Comment by ntrncic** [comment]
@ilcsik for packages you create a structure in the partners app. 
- Views
     - __init__.py
     - views.py
     - views_v2.py
- Serializers
    - __init__.py
    - serializers.py
    - serializers_v2.py

**Comment by ntrncic** [comment]
@ilcsik The reference number calculation appears to be the same as the document. yes it's only when agreement status changes to active.

**Comment by ilcsik** [comment]
Reference number calculation done in https://github.com/unicef/etools/pull/225/commits/f25ec647378c52ac469d68e3e910ba097d0a024c

**Comment by ilcsik** [comment]
Guys I have a couple of questions:
1. How do I get CountryProgramme for an Agreement the best way? There seems to be more route, and I'm not sure which one is the safest. It is needed for set CP's to_date on Agreement's end date in case of PCA on save.

2, Since reference_number is a property method I cannot include it in filtering/searching. The on save method seem to be putting reef number into agreement_number once it is present and save it, shall I use agreement number for searching/filtering based on reference number?

3, There's a requirement regarding changing status on save where all associated PD and SHPD should also be changed. What are those acronyms?

Thanks!

**Comment by ntrncic** [comment]
@ilcsik 
1. use CountryProgramme.current to get dates
2. use the agreement_number field for searching/filtering
3. Yes if status is changed to Suspended/Terminated the interventions (PCA model) associated to the agreement would also be suspended/terminated.

**Comment by ilcsik** [comment]
1, Done in https://github.com/unicef/etools/pull/225/commits/22542317eb830dd0b9f4979aaf018a3ef52f4ca3

2, Done in https://github.com/unicef/etools/pull/225/commits/da465d86b40f0a93c3af78176effff99a324be5c

3, Done in https://github.com/unicef/etools/pull/225/commits/41fa182bf4612a7518e31714038ef31be5473eec

There's a "cancelled" status on the PCA model, is that still valid now that we have suspended and terminated there?

**Comment by ilcsik** [comment]
Packages for API versions done in https://github.com/unicef/etools/pull/225/commits/37af855240577d4f2cf8f5d077dd8f40c3a54823

File: EquiTrack/partners/serializers/serializers_v2.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        )
+        read_only_fields = ("id",)
+
+
+class AgreementSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = Agreement
+        fields = (
+            "id",
```
Comment (ntrncic) [inline EquiTrack/partners/serializers/serializers_v2.py]: add partner name.

Comment (ilcsik) [inline EquiTrack/partners/serializers/serializers_v2.py]: Done in https://github.com/unicef/etools/pull/225/commits/61ed1f1bc2a4f65c7a20e96dd87811d11743b093

File: EquiTrack/partners/models.py [hunk 12]
```diff
@@ -716,9 +730,32 @@ def reference_number(self):
     def save(self, **kwargs):
 
         # commit the reference number to the database once the agreement is signed
-        if self.signed_by_unicef_date and not self.agreement_number:
+        if self.status != Agreement.DRAFT and self.signed_by_unicef_date and not self.agreement_number:
             self.agreement_number = self.reference_number
 
+        if self.signed_by_unicef_date and self.signed_by_partner_date:
+            start = max(self.signed_by_unicef_date, self.signed_by_partner_date)
+            self.start = start
```
Comment (robertavram) [inline EquiTrack/partners/models.py]: this should not be overridden on save anymore.. let's take this out.. it becomes part of validation instead

File: EquiTrack/partners/models.py [hunk 13]
```diff
@@ -716,9 +730,32 @@ def reference_number(self):
...
+                self.end = cp.to_date
+
+        if self.status in [Agreement.SUSPENDED, Agreement.TERMINATED]:
+            interventions = PCA.objects.filter(
+                                partner_id=self.partner.id,
+                                agreement_id=self.id,
+                                partnership_type__in=[PCA.PD, PCA.SHPD])
+            for item in interventions:
+                item.status = self.status
+                item.save()
```
Comment (robertavram) [inline EquiTrack/partners/models.py]: this piece of code will be transactional...  super.save should be called in the same transaction...

File: EquiTrack/partners/serializers/serializers.py [hunk 14]
```diff
@@ -260,6 +260,29 @@ class AgreementSerializer(serializers.ModelSerializer):
...
+            "bank_name",
+            "bank_address",
+            "account_title",
+            "account_number",
+            "routing_details",
+            "bank_contact_person",
+            "year",
+            "reference_number",
+        )
+        read_only_fields = ("id",)
```
Comment (robertavram) [inline EquiTrack/partners/serializers/serializers.py]: do we need to specify id as read-only? i thought drf auto-magically handles that

Comment (ilcsik) [inline EquiTrack/partners/serializers/serializers.py]: You're right, removed in https://github.com/unicef/etools/pull/225/commits/441d7db7661ecc4e8d414e51489eb7a372958599

File: EquiTrack/partners/models.py (hunk 13 above)
Comment (ilcsik) [inline EquiTrack/partners/models.py]: Indeed, done in https://github.com/unicef/etools/pull/225/commits/c971ef2775bc0d579a7e572d2c0e6ecfcec68c22

File: EquiTrack/partners/serializers/serializers_v2.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            "partner_manager",
+            "status",
+            "bank_name",
+            "bank_address",
+            "account_title",
+            "account_number",
+            "routing_details",
+            "bank_contact_person",
+            "year",
+            "reference_number",
```
Comment (ntrncic) [inline EquiTrack/partners/serializers/serializers_v2.py]: remove all bank details

File: EquiTrack/partners/urls_v2.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+    AgreementListAPIView,
+    AgreementDetailAPIView,
+    AgreementInterventionsListAPIView,
+)
+
+urlpatterns = (
+    url(r'^agreements/$', view=AgreementListAPIView.as_view(), name='agreement-list'),
+    url(r'^agreements/(?P<pk>\d+)/$', view=AgreementDetailAPIView.as_view(), name='agreement-detail'),
+    url(r'^partners/(?P<partner_pk>\d+)/agreements/$', view=AgreementListAPIView.as_view(), name='parter-agreement-list'),
+    url(r'^partners/(?P<partner_pk>\d+)/agreements/(?P<pk>\d+)/$', view=AgreementDetailAPIView.as_view(), name='partner-agreement-detail'),
```
Comment (ntrncic) [inline EquiTrack/partners/urls_v2.py]: this route is not needed for post, patch, delete requests

Comment (ntrncic) [inline EquiTrack/partners/urls_v2.py]: need agreements/<pk>/interventions

Comment (ilcsik) [inline EquiTrack/partners/urls_v2.py]: 1, I mean, the whole route is not needed then since pk is always unique so it does not make sense to filter based on partner id AND pk right?

2, Interventions added in https://github.com/unicef/etools/pull/225/commits/172c3400cc0168a970e30831d52d1e9e202c62bf

File: EquiTrack/partners/serializers/serializers_v2.py (hunk 15 above)
Comment (ilcsik) [inline EquiTrack/partners/serializers/serializers_v2.py]: 1, Is partner authorized officers list basically an M2M between Agreement and PartnerStaffMember model?

2, Bank details removed in https://github.com/unicef/etools/pull/225/commits/c5d2b49f8b4822e81dc2d765db724eab1e156efb

File: EquiTrack/partners/urls_v2.py [hunk 17]
```diff
@@ -9,6 +9,7 @@
 urlpatterns = (
     url(r'^agreements/$', view=AgreementListAPIView.as_view(), name='agreement-list'),
     url(r'^agreements/(?P<pk>\d+)/$', view=AgreementDetailAPIView.as_view(), name='agreement-detail'),
+    url(r'^agreements/(?P<pk>\d+)/interventions/$', view=AgreementInterventionsListAPIView.as_view(), name='agreement-interventions-list'),
     url(r'^partners/(?P<partner_pk>\d+)/agreements/$', view=AgreementListAPIView.as_view(), name='parter-agreement-list'),
     url(r'^partners/(?P<partner_pk>\d+)/agreements/(?P<pk>\d+)/$', view=AgreementDetailAPIView.as_view(), name='partner-agreement-detail'),
```
Comment (ntrncic) [inline EquiTrack/partners/urls_v2.py]: we don't need this route for agreement detail

**Comment by robertavram** [comment]
@ilcsik please update from develop

File: EquiTrack/partners/views/views_v2.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+
+class AgreementListAPIView(ListCreateAPIView):
+    """
+    Create new Agreements.
+    Returns a list of Agreements.
+    Returns related Interventions.
+    """
+    queryset = Agreement.objects.all()
+    serializer_class = AgreementSerializer
+    permission_classes = (PartnerManagerPermission,)
```
Comment (robertavram) [inline EquiTrack/partners/views/views_v2.py]: partner should not have access to this.. edit permission is valid only for usersin the partnership manager group... that is a group of unicef users... if you have any questions about these groups let me know

File: EquiTrack/partners/views/views_v2.py [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+            expression = functools.reduce(operator.and_, queries)
+            return q.filter(expression)
+        else:
+            return q.all()
+
+
+class AgreementInterventionsListAPIView(ListAPIView):
+    queryset = PCA.objects.all()
+    serializer_class = InterventionSerializer
+    permission_classes = (PartnerManagerPermission,)
```
Comment (robertavram) [inline EquiTrack/partners/views/views_v2.py]: same with interventions... go through all code and make sure the partners do not have access to this.
refer to the google doc for permission matrix... the top row represents user groups

File: EquiTrack/partners/serializers/serializers_v2.py (hunk 15 above)
Comment (ilcsik) [inline EquiTrack/partners/serializers/serializers_v2.py]: Ok 1. should be done thorugh partner - partnerstaffmember

**Comment by ilcsik** [comment]
Guys in regard of AuthorizedOfficers list on Agreement.

AuthorizedOfficers are created on Agreements save from the PartnerStaffMember that is added newly to the Agreement, so create is covered by that.

On the Agreement list serializer, we can have "authorized_officers" related name as a set of IDs

The requirements sheet says its a multiselect dropdown, so I think this is not enough. How should this work in regards of UI?

1, How do we suppose to add new AuthorizedOfficer instances? By a separate API call for AuthorizedOfficer (for eg like github assignees)? Or sending in a list of PartnerStaffMember IDs to the Agreement detail and handling it with custom save method (creating AuthorizedOfficer instances there)?

2, How do we suppose to list the available PartnerStaffMembers on the UI (on the admin this is handled by ChainedForeignKey I suppose)? Filter by partner on the PartnerStaffMembers list API?

3, How do we suppose to list the assigned AuthorizedOfficers on the Agreement detail? Extract name and ID from AuthorizedOfficer.officer and make a list on the serizlier?

Thanks

File: EquiTrack/partners/views/views_v2.py (hunk 18 above)
Comment (ilcsik) [inline EquiTrack/partners/views/views_v2.py]: @robertavram So the only user that can edit is the Agreement.partner_manager (which is an FK to PartnerStaffMember) right? Is that what you mean by "partnership manager group"?

Comment (ntrncic) [inline EquiTrack/partners/views/views_v2.py]: @ilcsik there are User Groups in Django and one of them is Partnership Manager. Check in the admin.

**Comment by robertavram** [comment]
Guys in regard of AuthorizedOfficers list on Agreement.

AuthorizedOfficers are created on Agreements save from the PartnerStaffMember that is added newly to the Agreement, so create is covered by that.
**True,**

On the Agreement list serializer, we can have "authorized_officers" related name as a set of IDs
**We don;t need this on the list**

The requirements sheet says its a multiselect dropdown, so I think this is not enough. How should this work in regards of UI?
**Not sure what you mean, multiselect dropdown on a list of partner staff members for that partner should be enough**

1, How do we suppose to add new AuthorizedOfficer instances? By a separate API call for AuthorizedOfficer (for eg like github assignees)? Or sending in a list of PartnerStaffMember IDs to the Agreement detail and handling it with custom save method (creating AuthorizedOfficer instances there)?
**Both ways eventually. Prioritize through the agreement form first we'll add the individual submission later**

2, How do we suppose to list the available PartnerStaffMembers on the UI (on the admin this is handled by ChainedForeignKey I suppose)? Filter by partner on the PartnerStaffMembers list API?

**We should have it in one of two ways: 1. The staff members (id + Full Name) is serialized on the Partner Org, or through an API /partners/pk/staff_members/ we should have both options available.**

3, How do we suppose to list the assigned AuthorizedOfficers on the Agreement detail? Extract name and ID from AuthorizedOfficer.officer and make a list on the serizlier?

**My approach to solve all this would be to add a ManyToMany field to Agreements on PartnerStaffMember for authorized on which agreements. Then write a script to migrate the existing data. and change in admin.**




Thanks

**Comment by ilcsik** [comment]
@ntrncic  when I'm doing migrations normalization on this branch, my makemigrations run generated a couple of migrations other than mine, as far as I see this is for your branch that I merged recently, is it ok to have them here?

Migrations for 'partners':
  0002_auto_20161119_0126.py:
    - Change managers on agreement
    - Add field status to agreement
    - Alter field status on agreementamendmentlog
    - Alter field status on amendmentlog
    - Alter field status on pca

**Comment by ilcsik** [comment]
@robertavram Thanks, I see. Let me recap what I'm doing, just to make sure we're on the same page and I get everything right.

1, We need to list PartnerStaffMembers that are available for an Agreement to be selected as AuthorizedOfficer. 

This is a PartnerStaffMember list filtered by PartnerOrganization (for that, Agreement.partner is used).

Two endpoint is needed:
GET /partners/pk/staff-members, 
GET /partners/pk, with embedded (id, fullname) list


2, We need to add new AuthorizedOfficers to a given Agreement, these are PartnerStaffMember ids selected from the list from point 1.

This is for now to be handled by sending in the id list to the Agreement endpoints (both create and update). 

POST /agreements
PUT/PATCH /agreements/pk

To implement this we either override create/update methods to add custom logic for creating/updating AuthorizedOfficers, OR define a M2M field with AuthorizedOfficer table as the through table, so that we can utilize RelatedManager. 

This way I think we can avoid data migration in both cases since we re-using AuthorizedOfficers as through table. Although admin form needs to be changed if that is to be used.


3, We need to have the Agreement's AuthorizedOfficers listed on the Agreement's detail view (but not on the list view). I suppose ID list will do here since this will always be matched with the available PartnerStaffMembers on the frontend, and the ones in the Agreements will be highlighted as selected.

GET /agreements/pk, with embedded ID list


I'm implementing all this on the partnerstaff-pr since this is mostly partnerstaff functionality and we already have the v2 api there and it is based on this branch.

Do you have anything to add? Thanks.

**Comment by ntrncic** [comment]
@ilcsik umm the amendment migrations shouldn't be merged. which branch did you merge from?

**Comment by ilcsik** [comment]
@ntrncic I pulled in agreements-api-v2 when you indicated

**Comment by ilcsik** [comment]
Just a note, as discussed on slack:

1, Remove AuthorizedOfficer model and use M2M between Agreement and PartnerStaffMember
2, Send fullname as well on the Agreements detail along with the ID for the PartnerStaffMember
